### PR TITLE
[Android] Load the game banner from the ROM and use it if no screenshot is available

### PR DESCRIPTION
**Pull request by mahdihijazi**
It doesn't look that good since the banners are 96x32 but it still better
than nothing.

### CODE REVIEW

**Comment by degasus** [comment]
Does this also work for Wii games without a same game? GC games ship a banner, Wii games doesn't. So we use the banner of the save game.

**Comment by mahdihijazi** [comment]
strange, because I see on the TV at least 2 banners, one for Mario Kart and another one for Rayman Origins. 

Can you tell me what is the "banner of the save game."? do you mean the saved screenshot?

**Comment by JosJuice** [comment]
By save game, degasus means the save data that the game stores in the Wii's NAND memory. Save games for Wii games contain a banner that's shown when viewing save data game in the Data Management screen in the Wii Menu, and this banner will be read by Dolphin when calling the C++ function GetBanner() (but only if there is a save game).

**Comment by JosJuice** [comment]
Also, semi-related: Wii banners are actually 192x64, but GetBanner() (in MainAndroid.cpp) mangles them down to 96x32 for no good reason.

**Comment by BhaaLseN** [comment]
Those pieces of code in `GameAdapter.java` and `GameRowPresenter.java` look awfully similar. Can you move that to a common location and reuse/call it from both, or is that not possible for reasons I don't know?

**Comment by mahdihijazi** [comment]
@JosJuice that explains everything. Then the answer to @degasus question is yes it works for wii games even if they don't provide the image, but it shows black image instead of the default one because GetBanner always return a color vector. I can't tell from the result if the banner is available or not.

@JosJuice do we have someone that can fix this from the cpp side, it would be better if we can show 192x64 banner, I can make any related change on the java side if needed.

**Comment by mahdihijazi** [comment]
@BhaaLseN I will try to create some PicassoUtils for this

**Comment by JosJuice** [comment]
Well... There are many people who can do the C++ part and a few that can do the Java part, but I don't know anyone that wants to do the JNI change we need so that the banner size can be returned along with the banner. I'll probably take a look at it sometime if nobody does it before me, but it might take quite a while before I do.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/RomBannerRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import com.squareup.picasso.Picasso;
+import com.squareup.picasso.Request;
+import com.squareup.picasso.RequestHandler;
+
+import org.dolphinemu.dolphinemu.NativeLibrary;
+
+import java.io.IOException;
+import java.nio.IntBuffer;
+
+public class RomBannerRequestHandler extends RequestHandler {
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/RomBannerRequestHandler.java]: Could we use a different name than `RomBannerRequestHandler`? We essentially never use the word ROMs in Dolphin to refer to games or disc images (but we do use it for things like the DSP ROMs). What does the rest of the Android code use? "Game"?

Comment (mahdihijazi) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/RomBannerRequestHandler.java]: GameBannerRequestHandler? it doesn't use anything else other than loading the banner from the ISO file

Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/RomBannerRequestHandler.java]: Sure.

**Comment by mahdihijazi** [comment]
all notes squashed back into the original commit

**Comment by degasus** [comment]
LGTM. @JosJuice I think we can merge this as it is. The JNI usage is well abstracted and easy to update.

**Comment by JosJuice** [comment]
Yeah, let's not block this PR on the Wii banner resolution problem.

**Comment by JosJuice** [comment]
We've been getting reports about this making Dolphin crash: https://forums.dolphin-emu.org/Thread-dolphin-crashes-when-going-to-wii-tab?pid=441674#pid441674
And it seems like the banner colors are wrong too: https://github.com/dolphin-emu/dolphin/pull/5265

Should we revert this merge for now, or do you know how to fix the crashing?

**Comment by mahdihijazi** [comment]
@JosJuice I will check it, thanks for the followup

**Pull request merged**